### ISSUE TITLE
Types not correctly inferred when using transformers (v3)

### ISSUE THREAD

**Issue by larsbs** [colinhacks/zod#355]
I'm trying to achieve one thing, basically, coalescing all `null` values into `undefined` when receiving data from an external provider. Since it's really annoying to write all the code necessary in `zod` to understand that again and again:

```ts
const myString = z.string().optional().nullable().transform((v) => v ?? undefined)
```

I was trying to write a wapper to do that automatically. It looks like the following:

```ts
function _optional<T>(zod: z.ZodType<T>): z.ZodTransformer<z.ZodNullable<z.ZodOptional<z.ZodType<T>>>, T | undefined> {
  return zod.optional().nullable().transform((v) => v ?? undefined);
}
```

However, this works well for simple types (for example, the one above):

```ts
const myString = _optional(ts.string());

// this gets correctly inferred as `string | undefined` instead of `string | null | undefined`
type t = z.infer<typeof myString>;
```

But it doesn't work for complex types like objects:

```ts
const UserSchema = _optional(z.object({
  email: _optional(z.string()),
  name: _optional(z.string()),
}));

type User = z.infer<typeof UserSchema>;
```

In the example above, the type `User` gets incorrectly inferred as: 

```ts
// notice the extra null typings
type User  = {
  email: string | null | undefined;
  name: string | null | undefined;
} | undefined
```

instead of 

```ts
type User  = {
  email: string | undefined;
  name: string | undefined;
} | undefined
```

like in the simple example at the beginning.

### Solutions

I could see two solutions, first one, see why this errors is happening (I don't know enough about zod internals as to answer this question myself just yet) or the second one, try to provide a simpler/nicer API for when you need some sort of default. For example:

```ts
const myString = z.string('default value');

const UserSchema = z.object({
  email: z.string(),
  name: z.string(),
}, { email: 'default email', name: 'default email' });
```

Basically meaning that every `zod` validator would accept an extra parameter that would be the default value in case the validated value is `null` or `undefined` (using `??`) and if that parameter is present, mark the validator automatically as `.optional().nullable()`

**Comment by colinhacks** [comment]
You were really close to the solution here. Writing generic functions on top of Zod is hard - I've been meaning to write up a guide describing the best practices.

Here's a correct implementation of `_optional`:

```ts
function _optional<T extends z.ZodTypeAny>(zod: T): z.ZodEffects<z.ZodNullableType<z.ZodOptionalType<T>>, T['_output'] | undefined> {
  return zod.optional().nullable().transform((v) => v ?? undefined) as any;
}
```

A few changes: 

1. The generic parameter should be z.ZodTypeAny (an alias for `z.ZodType<any,any,any>`). This let's TypeScript infer the subclass. If you do `_optional<T>(zod: z.ZodType<T>)` then the output type is always `ZodType` (the base class).
2. When you call `X.optional()` on a schema, it doesn't necessarily return `ZodOptional<X>`. If X is already a `ZodOptional` instance, it returns X instead of redundantly nesting two ZodOptionals. This is reflected in the `z.ZodOptionalType` alias - if you look at the type definition for the `.optional` method you'll see it. The same thing applies for nullables.
3. I couldn't get this working without casting the return type with `as any`. This is common (there are a lot of unavoidable `any`s in Zod's codebase). In this case you're smarter than TypeScript's inference engine so casting is required.

**Comment by larsbs** [comment]
Thanks a lot @colinhacks I always try to avoid `any`s in any codebase but I guess it's unavoidable in this case.

As I'm seeing in your solution, there's stuff (like `ZodOptionalType`, or `ZodNullableType`) that apparently requires some internal knowledge about zod, so yeah, talking a bit about this internal/helpers stuff would just help make this library even better (IMO, is the best right now if you're doing TS validation). I think something like this simple use case makes for a good scenario to start talking about this stuff.

**Comment by colinhacks** [comment]
Using `any` is really just a shortcut to stop TypeScript from complaining about the return type. You could do this without any like this:

```ts
function _optional<T extends z.ZodTypeAny>(zod: T) {
  return zod
    .optional()
    .nullable()
    .transform((v) => v ?? undefined) as z.ZodEffects<
    z.ZodNullableType<z.ZodOptionalType<T>>,
    T["_output"] | undefined
  >;
}
```

But I think this doesn't look as nice.

But in general being dogmatically opposed to `any` is untenable when you're doing complex things. For instance, it's necessary when doing inference over classes with constrained generic parameters. For instance this doesn't work:

<img width="633" alt="Screen Shot 2021-03-24 at 10 46 16 AM" src="https://user-images.githubusercontent.com/3084745/112349267-31949d80-8c8e-11eb-8025-c01690e172d8.png">

But using `extends ZodType<any,any,any>` does. 

I'm thinking about getting rid of ZodOptionalType and ZodNullableType for the reasons you describe. I'll try to add a section on generics (or at least link to this issue from the README).

**Comment by larsbs** [comment]
@colinhacks all is clear on my side, you can close the issue or leave it open, whatever you usually do. Thanks.

**Comment by ElhanM** [comment]
The as z.ZodEffects<...> assertion ensures that after transformation, the type is strictly the transformed type (Date or Prisma.Decimal), not any of the original types.

```
// Helper schema for DateTime fields
const dateSchema = z
  .instanceof(Date)
  .or(z.string())
  .refine((value) => {
    try {
      return !isNaN(new Date(value).getTime())
    } catch (error) {
      return false
    }
  })
  .transform((value) => new Date(value)) as z.ZodEffects<
    z.ZodEffects<
      z.ZodType<Date, z.ZodTypeDef, Date>,
      Date,
      Date
    >,
    Date,
    Date
  >

// Helper schema for Decimal fields
const decimalSchema = z
  .instanceof(Prisma.Decimal)
  .or(z.string())
  .or(z.number())
  .refine((value) => {
    try {
      return new Prisma.Decimal(value)
    } catch (error) {
      return false
    }
  })
  .transform((value) => new Prisma.Decimal(value)) as z.ZodEffects<
    z.ZodEffects<
      z.ZodType<Prisma.Decimal, z.ZodTypeDef, Prisma.Decimal>,
      Prisma.Decimal,
      Prisma.Decimal
    >
  >
```
